src/controller/OperationalCredentialsDelegate.h

```diff
@@ -36,24 +41,27 @@ class DLL_EXPORT OperationalCredentialsDelegate
 
     /**
      * @brief
-     *   This function generates an operational certificate for the given node.
+     *   This function generates an operational certificate for the device.
      *   The API generates the certificate in X.509 DER format.
      *
      *   The delegate is expected to use the certificate authority whose certificate
      *   is returned in `GetIntermediateCACertificate()` or `GetRootCACertificate()`
      *   API calls.
      *
-     * @param[in] peerId       Node ID and Fabric ID of the target device.
-     * @param[in] csr          Certificate Signing Request from the node in DER format.
-     * @param[in] serialNumber Serial number to assign to the new certificate.
-     * @param[in] certBuf      The API will fill in the generated cert in this buffer. The buffer is allocated by the caller.
-     * @param[in] certBufSize  The size of certBuf buffer.
-     * @param[out] outCertLen  The size of the actual certificate that was written in the certBuf.
+     *   The delegate will call `onNOCGenerated` when the NOC is ready.
+     *
+     * @param[in] nodeId          Optional node ID. If provided, the generated NOC must use the provided ID.
+     *                            If ID is not provided, the delegate must generate one.
+     * @param[in] fabricId        Fabric ID for which the certificate is being requested.
+     * @param[in] csr             Certificate Signing Request from the node in DER format.
+     * @param[in] DAC             Device attestation certificate received from the device being commissioned
+     * @param[in] onNOCGenerated  Callback handler to provide generated NOC to the caller of GenerateNodeOperationalCertificate()
      *
      * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
      */
-    virtual CHIP_ERROR GenerateNodeOperationalCertificate(const PeerId & peerId, const ByteSpan & csr, int64_t serialNumber,
-                                                          uint8_t * certBuf, uint32_t certBufSize, uint32_t & outCertLen) = 0;
+    virtual CHIP_ERROR GenerateNodeOperationalCertificate(const Optional<NodeId> & nodeId, FabricId fabricId, const ByteSpan & csr,
```

Comment:
I'm approving as this PR is needed to unblock some activities.

However, I want to point out that even with NodeID as an optional parameter, this interface misses the mark.  There should be no concept under `project-chip/connectedhomeip/src` of the commissioner proposing a node ID, or even fabric ID, to the CA infrastructure.  That is a possible implementation.  It is equally possible that all node IDs will be assigned by the CA itself.

The right approach is for such local assignment of IDs in our example implementations to occur in code under `project-chip/connectedhomeip/examples`.  The interface here must not specify IDs; instead, an opaque context should be passed from the app.  There will be an implicit contract between app and ecosystem-specific CA infrastructure as to what context is needed, but it is impossible for any code under `project-chip/connectedhomeip/src` to know or predict what that will be.

The immediate consequence here is that any CA-assignment of a node ID or fabric ID other than what is passed is likely to break the commissioner.  I don't think it's overstating the problem to call this a bug.   And until it's fixed, any implementation that adopts an approach of CA-assignment of all IDs will have to fork and patch the sdk to work around the problem.

Reply:
@msandstedt these are the main reasons we have the optional node ID in the current CA interface.

1. In the current commissioner interface, the node ID is provided by the commissioner app when the `PairDevice()` is called. I understand that it's not optimal and may even be incorrect. But unless we update all current test tools and apps as part of this PR, we will end up breaking the tip of the tree. Given we are using test events as project schedule, and TE4 cut off is 1 week away, not sure if we want to introduce all those changes in this PR.
2. `UpdateNOC` command will update the existing NOC on a device with the new NOC. Would node ID be updated as part of this process? I tried reading the spec and could not conclude one way or another. But in case we want to retain the same node ID as part of `UpdateNOC`, this optional node ID parameter can be used to inform CA of the requested node ID value. Again, may be it is not needed, or there are other ways of achieving it. Just my thoughts..

If there is no use case where the optional node ID is needed, it can certainly be removed from the interface. Removing it right now will just cause the code to break.

The commissioner interface was designed before the specifications were finalized, and it is evolving as features are getting integrated. I understand the problem an evolving public API can cause to the customers of the API. Capturing the API requirements in an issue and getting it prioritized in a test event schedule might be a good way to get to the final solution in a deterministic timeline.

Reply:
> Would node ID be updated as part of this process?

Yes, if the new certificate has a new node id.

> I tried reading the spec and could not conclude one way or another.

"Effect When Received" for UpdateNOC says:

> ... such that the Node’s Operational Identifier within the Fabric immediately changes.

and

> All internal data reflecting the prior operational identifier of the Node within the Fabric SHALL be revoked and removed, to an outcome equivalent to the disappearance of the prior Node, except for  ....

If there's something we can do to make the spec clearer here, we should do it...

Reply:
> 1. In the current commissioner interface, the node ID is provided by the commissioner app when the `PairDevice()` is called. I understand that it's not optimal and may even be incorrect. But unless we update all current test tools and apps as part of this PR, we will end up breaking the tip of the tree. Given we are using test events as project schedule, and TE4 cut off is 1 week away, not sure if we want to introduce all those changes in this PR.

The problem is that with controllers in scope for TE4, it is no longer sufficient for the sdk to only support the in-tree use cases.  The only alternative would be for test event participants to develop infrastructure solely for the purpose of test event compatibility.  That's not desirable.

I've filed https://github.com/project-chip/connectedhomeip/issues/7951 to track this.

Reply:
> > Would node ID be updated as part of this process?
> 
> Yes, if the new certificate has a new node id.
> 
> > I tried reading the spec and could not conclude one way or another.
> 
> "Effect When Received" for UpdateNOC says:
> 
> > ... such that the Node’s Operational Identifier within the Fabric immediately changes.
> 
> and
> 
> > All internal data reflecting the prior operational identifier of the Node within the Fabric SHALL be revoked and removed, to an outcome equivalent to the disappearance of the prior Node, except for  ....
> 
> If there's something we can do to make the spec clearer here, we should do it...

Is it desired/required for the node ID to be changed as part of `UpdateNOC` command? Maybe it's left up to the commissioner implementation. If the use case allows the commissioner to continue to use the same node ID for the device, having the Optional node ID in the CA interface would be useful.

Reply:
It's fine for a commissioner implementation to propose its node ID to ecosystem-specific signing infrastructure.  It's not appropriate to formalize that type of flow in the `chip::controller::DeviceCommissioner` object.

We need to invert our design here.  Ecosystem-specific signing infrastructure will require context to field CSRs that the sdk cannot know or predict.  Issues like https://github.com/project-chip/connectedhomeip/issues/6656 are well-meaning but misguided.  We do not know how, where or if ecosystems will embed other RDNs.  It makes no sense to formalize local APIs for this in the first place.

At the simplest level, app code should pass a `void *` context that will be completely opaque to the sdk and only discernible by the associated `OperationalCredentialsDelegate` implementation.  Consider this: how does ecosystem-specific signing infrastructure verify the commissioner's permission to commission a particular node?  `OperationalCredentialsDelegate` needs some context from the app to prove this permission.  What this should look like though is unknowable from the sdk perspective, just like fabric ID, node ID and tag assignment rules are unknowable from the sdk perspective.

Reply:
@pan-apple @bzbarsky-apple @msandstedt what are the next steps here?

Reply:
I think we all agree that the app is the only thing that has the full context here; the SDK does not.  So the SDK needs to  communicate some sort of app-meaningful context indicator here.

In addition to that, we should figure out whether there are _parts_ of the context that the SDK can provide to the app and if so whether that provision is useful in common cases.  That could include things like "the CSR is for a node with a certain id", or "the CSR is for a node that does not yet have a node id", or whatnot.  If we think the app would always have this information anyway, there's not much point communicating it, of course.

Reply:
Was merged before comments were resolved, I thought I enabled this - will investigate. I want to make sure that we have a plan moving forward here though as there is obviously clarification required.